ratelimits/limiter.go

```diff
@@ -325,15 +322,27 @@ func (l *Limiter) BatchSpend(ctx context.Context, txns []Transaction) (*Decision
 
 	if batchDecision.allowed {
 		if len(newBuckets) > 0 {
-			createdBuckets, err := l.source.BatchSetNotExisting(ctx, newBuckets)
+			// Use BatchSetNotExisting to initialize new buckets so that we
+			// detect if concurrent requests have created this bucket at the
+			// same time, which would result in overwriting if we used a plain
+			// "SET" command. If that happens, fall back to incrementing.
+			initializationResults, err := l.source.BatchSetNotExisting(ctx, newBuckets)
 			if err != nil {
 				return nil, fmt.Errorf("batch set for %d keys: %w", len(newBuckets), err)
 			}
-			for k, created := range createdBuckets {
-				if !created {
-					// A bucket was created by another request, fall back to
-					// BatchSet.
-					staleBuckets[k] = newBuckets[k]
+			existingBuckets := make(map[string]struct{})
+			for k, initialized := range initializationResults {
+				if !initialized {
+					existingBuckets[k] = struct{}{}
+				}
+			}
+			for _, v := range txns {
+				_, bucketExists := existingBuckets[v.bucketKey]
+				if bucketExists {
```

Comment:
The `map[X]struct{}` trick is a cool way to economize on storage space, but it's probably not necessary here. It complicates the code a bit, and differs from the style returned by `BatchSetNotExisting`, which is `map[string]bool`. If we use the same type here we get to simplify:

```suggestion
			existingBuckets := make(map[string]bool)
			for k, initialized := range initializationResults {
				if !initialized {
					existingBuckets[k] = true
				}
			}
			for _, v := range txns {
				if existingBuckets[v.bucketKey] {
```

But actually we can do one better than that: we can rely on the fact that `BatchSetNotExisting` is guaranteed to return a map entry for each input (which we should probably document) to skip all the code that sets up a second map:

```suggestion
			existingBuckets := make(map[string]struct{})
			for k, initialized := range initializationResults {
				if !initialized {
					incrBuckets[v.bucketKey] = increment{
						cost: time.Duration(v.cost * v.limit.emissionInterval),
						ttl:  time.Duration(v.limit.burstOffset),
					}
				}
			}
```

I think this may actually fix a bug: the code you have here ranges over `txns`, but we only called `BatchSetNotExisting` for `newBuckets`, which is a subset. I think the code as it is here would, for instance, cause items that are already in `staleBuckets` to get added to `incrBuckets` as well.

Also, optional API design tweak: What if `BatchSetNotExisting` returned instead a slice containing just the bucket keys that weren't created? Since that's the only information we really care about here, it follows the idea of "least mechanism" and is maybe a little clearer to reason about.

Reply:
Went over this live and Samantha's right, there wasn't a bug. We iterate over the list of original transactions but only pluck out the ones that we attempted to create (and failed due to already existing).

Also we worked out a slight rearrangement that may make this property more clear.